### PR TITLE
Revived google-readability-casting

### PR DESCRIPTION
**Pull request by veblush**
Revived `google-readability-casting` which was disabled temporarily by https://github.com/grpc/grpc/pull/31518. It's still acting up on some edge cases but it's easy to tame using brace initialization.

### CODE REVIEW

**Comment by ctiller** [comment]
Stating for the record that this seems like a readability regression to me: what's the upside?

File: src/core/ext/filters/client_channel/lb_policy/subchannel_list.h [hunk 1]
```diff
@@ -324,8 +324,8 @@ void SubchannelData<SubchannelListType,
   pending_watcher_ =
       new Watcher(this, subchannel_list()->WeakRef(DEBUG_LOCATION, "Watcher"));
   subchannel_->WatchConnectivityState(
-      std::unique_ptr<SubchannelInterface::ConnectivityStateWatcherInterface>(
-          pending_watcher_));
+      std::unique_ptr<SubchannelInterface::ConnectivityStateWatcherInterface>{
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/subchannel_list.h:327]: Why is this change necessary?  `std::unique_ptr<SubchannelInterface::ConnectivityStateWatcherInterface>` is a class type, so parens should be fine here.  This is not a cast, it's constructing a new object.

File: src/core/ext/xds/xds_listener.cc [hunk 2]
```diff
@@ -628,8 +628,8 @@ absl::optional<XdsListenerResource::FilterChainMap::CidrRange> CidrRangeParse(
         google_protobuf_UInt32Value_value(prefix_len_proto),
         (reinterpret_cast<const grpc_sockaddr*>(cidr_range.address.addr))
                     ->sa_family == GRPC_AF_INET
-            ? uint32_t(32)
-            : uint32_t(128));
+            ? static_cast<uint32_t>(32)
```
Comment (markdroth) [inline src/core/ext/xds/xds_listener.cc:631]: I think these two lines can use braces instead of `static_cast<>`.

File: src/core/lib/address_utils/sockaddr_utils.cc [hunk 3]
```diff
@@ -371,7 +371,7 @@ void grpc_sockaddr_mask_bits(grpc_resolved_address* address,
     } else if (mask_bits >= 32) {
       return;
     }
-    uint32_t mask_ip_addr = (~(uint32_t(0))) << (32 - mask_bits);
+    uint32_t mask_ip_addr = (~(static_cast<uint32_t>(0))) << (32 - mask_bits);
```
Comment (markdroth) [inline src/core/lib/address_utils/sockaddr_utils.cc:374]: All of the ocurrances in this file can use braces instead of `static_cast<>`.

File: src/core/lib/gpr/string.cc [hunk 4]
```diff
@@ -89,7 +89,7 @@ static dump_out dump_out_create(void) {
 
 static void dump_out_append(dump_out* out, char c) {
   if (out->length == out->capacity) {
-    out->capacity = std::max(size_t(8), 2 * out->capacity);
+    out->capacity = std::max(static_cast<size_t>(8), 2 * out->capacity);
```
Comment (markdroth) [inline src/core/lib/gpr/string.cc:92]: The occurrances in this file can use braces instead of `static_cast<>`.

File: src/core/lib/http/parser.cc [hunk 5]
```diff
@@ -337,7 +337,8 @@ static grpc_error_handle addbyte_body(grpc_http_parser* parser, uint8_t byte) {
   }
 
   if (*body_length == parser->body_capacity) {
-    parser->body_capacity = std::max(size_t(8), parser->body_capacity * 3 / 2);
+    parser->body_capacity =
+        std::max(static_cast<size_t>(8), parser->body_capacity * 3 / 2);
```
Comment (markdroth) [inline src/core/lib/http/parser.cc:341]: This should use braces.

File: src/core/lib/iomgr/ev_poll_posix.cc [hunk 6]
```diff
@@ -1193,7 +1193,7 @@ static void pollset_set_add_pollset(grpc_pollset_set* pollset_set,
   gpr_mu_lock(&pollset_set->mu);
   if (pollset_set->pollset_count == pollset_set->pollset_capacity) {
     pollset_set->pollset_capacity =
-        std::max(size_t(8), 2 * pollset_set->pollset_capacity);
+        std::max(static_cast<size_t>(8), 2 * pollset_set->pollset_capacity);
```
Comment (markdroth) [inline src/core/lib/iomgr/ev_poll_posix.cc:1196]: All the changes in this file should use braces.

File: src/core/lib/resource_quota/memory_quota.cc [hunk 7]
```diff
@@ -371,7 +371,8 @@ void BasicMemoryQuota::Start() {
           std::tuple<const char*, RefCountedPtr<ReclaimerQueue::Handle>> arg) {
         auto reclaimer = std::move(std::get<1>(arg));
         if (GRPC_TRACE_FLAG_ENABLED(grpc_resource_quota_trace)) {
-          double free = std::max(intptr_t(0), self->free_bytes_.load());
+          double free =
+              std::max(static_cast<intptr_t>(0), self->free_bytes_.load());
```
Comment (markdroth) [inline src/core/lib/resource_quota/memory_quota.cc:375]: The changes in this file should use braces.

I'll stop noting these individually, but please fix throughout the PR.  Anywhere that the argument is a constant value, it's probably better to use braces than call it a cast.

File: test/core/transport/binder/binder_transport_test.cc [hunk 8]
```diff
@@ -192,8 +192,8 @@ class MetadataEncoder {
   template <typename Which>
   void Encode(Which, const typename Which::ValueType& value) {
     metadata_.emplace_back(
-        std::string(Which::key()),
-        std::string(grpc_core::Slice(Which::Encode(value)).as_string_view()));
+        std::string{Which::key()},
```
Comment (markdroth) [inline test/core/transport/binder/binder_transport_test.cc:195]: `std::string` is a class, so this should not require braces.  It's constructing a new object, not doing a cast.

**Review by markdroth — COMMENTED** [review]
@ctiller The rationale for this check is in https://google.github.io/styleguide/cppguide.html#Casting.  I agree that the recent changes have made it a little more cumbersome, though.

I do think we need a check to avoid C-style casts, because those do actually cause bugs.  If there was another flag we could use here that would continue to do that check without any of the new requirements that they seem to have added here in the new version of clang, I'd be open to that.  But if we have to choose between enabling this the way it is or leaving it off, I'd rather enable it.

File: src/core/ext/filters/client_channel/lb_policy/subchannel_list.h (hunk 1 above)
Comment (veblush) [reply src/core/ext/filters/client_channel/lb_policy/subchannel_list.h:327]: This is to address this weird false positive of clang-tidy. Clang-tidy keeps saying it should change so I did it to workaround it.

File: test/core/transport/binder/binder_transport_test.cc (hunk 8 above)
Comment (veblush) [reply test/core/transport/binder/binder_transport_test.cc:195]: Like the above case, this is again to address this weird false positive of clang-tidy. There were two cases in total.

**Comment by veblush** [comment]
@markdroth To address your comments; I replaced `static_cast<(\w+)>\((\d+)\)` with `$1{$2}` throughout all gRPC code.

File: src/core/ext/filters/client_channel/lb_policy/subchannel_list.h (hunk 1 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy/subchannel_list.h:327]: This seems obviously wrong.

Can we instead leave it as-is and use a NOLINT comment to have clang-tidy leave it alone?

File: test/core/transport/binder/binder_transport_test.cc (hunk 8 above)
Comment (markdroth) [reply test/core/transport/binder/binder_transport_test.cc:195]: I'd prefer to use a NOLINT comment instead here.  I think using parens is correct here.

File: test/core/channel/channelz_test.cc [hunk 9]
```diff
@@ -449,16 +449,16 @@ TEST_F(ChannelzRegistryBasedTest, GetTopChannelsMoreGaps) {
   Json array = (*parsed_json->mutable_object())["channel"];
   ValidateJsonArraySize(array, 2);
   std::vector<intptr_t> uuids = GetUuidListFromArray(array.array_value());
-  EXPECT_EQ(static_cast<intptr_t>(3), uuids[0]);
-  EXPECT_EQ(static_cast<intptr_t>(5), uuids[1]);
+  EXPECT_EQ(intptr_t{3}, uuids[0]);
```
Comment (markdroth) [inline test/core/channel/channelz_test.cc:452]: In all of the cases where we're specifying a type inside of a gtest `EXPECT_*` macro, it's generally better to just use `U` or `UL` suffixes instead of spelling out the specific type.  So for example, in this case, you can change `intptr_t{3}` to `3UL`.  Or if a test currently says `unsigned{1}`, you can change that to `1U`.

Please fix throughout.

Comment (ctiller) [reply test/core/channel/channelz_test.cc:452]: Is 3UL guaranteed to be intptr_t on all platforms? (I genuinely don't know)

File: test/core/transport/binder/binder_transport_test.cc (hunk 8 above)
Comment (veblush) [reply test/core/transport/binder/binder_transport_test.cc:195]: I don't have a strong opinion here but shouldn't `std::string{}` be a proper construction call?

File: test/core/channel/channelz_test.cc (hunk 9 above)
Comment (markdroth) [reply test/core/channel/channelz_test.cc:452]: Good question.  I know that I've always used a `UL` suffix in `EXPECT_*` macros for things like `size_t`, and it's always worked fine, so I assume it will for `intptr_t` too.  But if it causes the test to fail on any platform we care about, we can always switch back to the current syntax for any `intptr_t` values.

File: test/core/transport/binder/binder_transport_test.cc (hunk 8 above)
Comment (markdroth) [reply test/core/transport/binder/binder_transport_test.cc:195]: I believe that calls the initializer_list ctor, not the single-argument ctor.  But there's no reason we should need to do that here.  The braces are ugly and confusing, and they really don't provide any benefit in this case.

File: test/core/channel/channelz_test.cc (hunk 9 above)
Comment (veblush) [reply test/core/channel/channelz_test.cc]: `EXPECT_EQ` doesn't require the type of two given value to have the exact same so I think it should be fine with just plain 3 here.

Comment (markdroth) [reply test/core/channel/channelz_test.cc]: Historically, I have found these suffixes to frequently be needed to get `EXPECT_*` macros to compile properly on all platforms.  But if it works without it, that's fine with me.

File: test/core/transport/binder/binder_transport_test.cc (hunk 8 above)
Comment (veblush) [reply test/core/transport/binder/binder_transport_test.cc]: This could be actually both a regular ctor and initializer_list ctor. Yay another C++ thing! But I'm adding NOLINT comment here.

File: src/core/ext/transport/chttp2/transport/flow_control.cc [hunk 10]
```diff
@@ -255,7 +255,8 @@ TransportFlowControl::TargetInitialWindowSizeBasedOnMemoryPressureAndBdp()
   //                                                                pressure
   const double kAnythingGoesPressure = 0.2;
   const double kAdjustedToBdpPressure = 0.5;
-  const double kAnythingGoesWindow = std::max(double(1 << 24), bdp);
+  const double kAnythingGoesWindow =
+      std::max(static_cast<double>(1 << 24), bdp);
```
Comment (markdroth) [inline src/core/ext/transport/chttp2/transport/flow_control.cc:259]: This one should probably also use braces instead of `static_cast<>`.

File: test/core/iomgr/buffer_list_test.cc [hunk 11]
```diff
@@ -47,12 +47,12 @@ TEST(BufferListTest, Testshutdownflusheslist) {
...
   gpr_atm verifier_called[NUM_ELEM];
   for (auto i = 0; i < NUM_ELEM; i++) {
-    gpr_atm_rel_store(&verifier_called[i], static_cast<gpr_atm>(0));
+    gpr_atm_rel_store(&verifier_called[i], gpr_atm{0});
     tb_list.AddNewEntry(i, 0, static_cast<void*>(&verifier_called[i]));
   }
   tb_list.Shutdown(nullptr, absl::OkStatus());
   for (auto i = 0; i < NUM_ELEM; i++) {
-    ASSERT_EQ(gpr_atm_acq_load(&verifier_called[i]), static_cast<gpr_atm>(1));
+    ASSERT_EQ(gpr_atm_acq_load(&verifier_called[i]), gpr_atm{1});
```
Comment (markdroth) [inline test/core/iomgr/buffer_list_test.cc:55]: This can presumably just say `1`, with no explicit type specified.

File: test/core/iomgr/buffer_list_test.cc [hunk 12]
```diff
@@ -82,10 +82,10 @@ TEST(BufferListTest, Testverifiercalledonack) {
       TestVerifierCalledOnAckVerifier);
   grpc_core::TracedBufferList tb_list;
   gpr_atm verifier_called;
-  gpr_atm_rel_store(&verifier_called, static_cast<gpr_atm>(0));
+  gpr_atm_rel_store(&verifier_called, gpr_atm{0});
   tb_list.AddNewEntry(213, 0, &verifier_called);
   tb_list.ProcessTimestamp(&serr, nullptr, &tss);
-  ASSERT_EQ(gpr_atm_acq_load(&verifier_called), static_cast<gpr_atm>(1));
+  ASSERT_EQ(gpr_atm_acq_load(&verifier_called), gpr_atm{1});
```
Comment (markdroth) [inline test/core/iomgr/buffer_list_test.cc:88]: Same here.

File: test/core/iomgr/buffer_list_test.cc [hunk 13]
```diff
@@ -102,10 +102,10 @@ TEST(BufferListTest, Testrepeatedshutdown) {
...
```
Comment (markdroth) [inline test/core/iomgr/buffer_list_test.cc:108]: And here.

File: test/core/transport/chttp2/context_list_test.cc [hunk 14]
```diff
@@ -97,13 +97,13 @@ TEST_F(ContextListTest, ExecuteFlushesList) {
...
     s[i]->byte_counter = kByteOffset;
-    gpr_atm_rel_store(&verifier_called[i], static_cast<gpr_atm>(0));
+    gpr_atm_rel_store(&verifier_called[i], gpr_atm{0});
     ContextList::Append(&list, s[i]);
   }
   Timestamps ts;
   ContextList::Execute(list, &ts, absl::OkStatus());
   for (auto i = 0; i < kNumElems; i++) {
-    EXPECT_EQ(gpr_atm_acq_load(&verifier_called[i]), static_cast<gpr_atm>(1));
+    EXPECT_EQ(gpr_atm_acq_load(&verifier_called[i]), gpr_atm{1});
```
Comment (markdroth) [inline test/core/transport/chttp2/context_list_test.cc:106]: And here.

File: test/core/transport/chttp2/context_list_test.cc [hunk 15]
```diff
@@ -151,12 +151,12 @@ TEST_F(ContextListTest, NonEmptyListEmptyTimestamp) {
...
     s[i]->context = &verifier_called[i];
     s[i]->byte_counter = kByteOffset;
-    gpr_atm_rel_store(&verifier_called[i], static_cast<gpr_atm>(0));
+    gpr_atm_rel_store(&verifier_called[i], gpr_atm{0});
     ContextList::Append(&list, s[i]);
   }
   ContextList::Execute(list, nullptr, absl::OkStatus());
   for (auto i = 0; i < kNumElems; i++) {
-    EXPECT_EQ(gpr_atm_acq_load(&verifier_called[i]), static_cast<gpr_atm>(1));
+    EXPECT_EQ(gpr_atm_acq_load(&verifier_called[i]), gpr_atm{1});
```
Comment (markdroth) [inline test/core/transport/chttp2/context_list_test.cc:159]: And here.

File: test/cpp/common/channel_arguments_test.cc [hunk 16]
```diff
@@ -99,7 +99,7 @@ class ChannelArgumentsTest : public ::testing::Test {
   void VerifyDefaultChannelArgs() {
     grpc_channel_args args;
     SetChannelArgs(channel_args_, &args);
-    EXPECT_EQ(static_cast<size_t>(1), args.num_args);
+    EXPECT_EQ(size_t{1}, args.num_args);
```
Comment (markdroth) [inline test/cpp/common/channel_arguments_test.cc:102]: And here.

File: test/cpp/util/byte_buffer_test.cc [hunk 17]
```diff
@@ -70,7 +70,7 @@ TEST_F(ByteBufferTest, Clear) {
   Slice s(kContent1);
   ByteBuffer buffer(&s, 1);
   buffer.Clear();
-  EXPECT_EQ(static_cast<size_t>(0), buffer.Length());
+  EXPECT_EQ(size_t{0}, buffer.Length());
```
Comment (markdroth) [inline test/cpp/util/byte_buffer_test.cc:73]: And here.

**Comment by veblush** [comment]
@markdroth I addressed your last comments (thanks!) and I had to rebase this PR onto master to address one clang-tidy error introduced by a recent commit.

**Comment by markdroth** [comment]
Please do not rebase once a review has started!  That makes it basically impossible for the reviewer to see what has changed since their last review.

Instead, please just merge master into your branch and push the merge commit.

**Review by markdroth — APPROVED** [review]
